internal/service/managedobjectstorage/managed_object_storage.go

```diff
@@ -0,0 +1,360 @@
+				Optional:    true,
+				Type:        schema.TypeSet,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+		},
+	}
+}
+
+func schemaEndpoint() *schema.Resource {
+	return &schema.Resource{
+		Description: "OpenSearch index",
+		Schema: map[string]*schema.Schema{
+			"domain_name": {
+				Description: "Domain name of the endpoint.",
+				Computed:    true,
+				Type:        schema.TypeString,
+			},
+			"type": {
+				Description: "Type of the endpoint (`private` / `public`).",
+				Computed:    true,
+				Type:        schema.TypeString,
+			},
+		},
+	}
+}
+
+func schemaNetwork() *schema.Resource {
+	return &schema.Resource{
+		Description: "Managed Object Storage network",
+		Schema: map[string]*schema.Schema{
+			"family": {
+				Description: "Network family. IPv6 currently not supported.",
+				Required:    true,
+				Type:        schema.TypeString,
+				ValidateFunc: validation.StringInSlice(
+					[]string{
+						upcloud.IPAddressFamilyIPv4,
+						// IPv6 currently not supported
+					},
+					false,
+				),
+			},
+			"name": {
+				Description: "Network name. Must be unique within the service.",
+				Required:    true,
+				Type:        schema.TypeString,
+				ValidateFunc: validation.All(
+					validation.StringLenBetween(1, 64),
+					validation.StringMatch(regexp.MustCompile(`^[a-zA-Z0-9_-]+$`), ""),
+				),
+			},
+			"type": {
+				Description: "Network type.",
+				Required:    true,
+				Type:        schema.TypeString,
+				ValidateFunc: validation.StringInSlice(
+					[]string{
+						"private",
+						"public",
+					},
+					false,
+				),
+			},
+			"uuid": {
+				Description: "Private network uuid. For public networks the field should be omitted.",
+				Optional:    true,
+				Type:        schema.TypeString,
+			},
+		},
+	}
+}
+
+func resourceManagedObjectStorageCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) (diags diag.Diagnostics) {
+	svc := meta.(*service.Service)
+
+	req := &request.CreateManagedObjectStorageRequest{
+		ConfiguredStatus: upcloud.ManagedObjectStorageConfiguredStatus(d.Get("configured_status").(string)),
+		Region:           d.Get("region").(string),
+	}
+
+	if v, ok := d.GetOk("labels"); ok {
+		req.Labels = utils.LabelsMapToSlice(v.(map[string]interface{}))
+	}
+
+	networks, err := networksFromResourceData(d)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	req.Networks = networks
+
+	userReqs := make([]request.ManagedObjectStorageUser, 0)
+	users, ok := d.GetOk("users")
+	if ok {
+		for _, user := range users.(*schema.Set).List() {
+			userReqs = append(req.Users, request.ManagedObjectStorageUser{
+				Username: user.(string),
+			})
+		}
+	}
+	req.Users = userReqs
+
+	storage, err := svc.CreateManagedObjectStorage(ctx, req)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.SetId(storage.UUID)
+
+	waitReq := &request.WaitForManagedObjectStorageOperationalStateRequest{
```

Comment:
Not related to this PR, but should we pass context instead of timeout to these `WaitFor*State` functions?

Reply:
Good point, ticket created.